tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc

```diff
@@ -0,0 +1,369 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.h"
+
+#include <functional>
+#include <string>
+
+#include "tensorflow/compiler/xla/hlo/ir/hlo_casting_utils.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_instruction.h"
+#include "tensorflow/compiler/xla/literal_util.h"
+#include "tensorflow/compiler/xla/primitive_util.h"
+#include "tensorflow/compiler/xla/service/gpu/backend_configs.pb.h"
+#include "tensorflow/compiler/xla/service/gpu/cublas_cudnn.h"
+#include "tensorflow/compiler/xla/service/gpu/matmul_utils.h"
+#include "tensorflow/compiler/xla/service/hlo_creation_utils.h"
+#include "tensorflow/compiler/xla/service/pattern_matcher.h"
+#include "tensorflow/compiler/xla/stream_executor/dnn.pb.h"
+#include "tensorflow/core/platform/errors.h"
+#include "tensorflow/core/platform/statusor.h"
+#include "tensorflow/core/platform/stream_executor_no_cuda.h"
+
+namespace xla {
+namespace gpu {
+namespace {
+namespace m = match;
+
+bool IsFMHACustomCall(const HloInstruction* instr) {
+  return instr->opcode() == HloOpcode::kCustomCall &&
+         (instr->custom_call_target() == kCudnnfMHABmmBmmCallTarget ||
+          instr->custom_call_target() ==
+              kCudnnfMHAScaleBiasMaskSoftmaxCallTarget ||
+          instr->custom_call_target() ==
+              kCudnnfMHAScaleBiasMaskSoftmaxDropoutCallTarget ||
+          instr->custom_call_target() == kCudnnfMHAScaleMaskSoftmaxCallTarget ||
+          instr->custom_call_target() ==
+              kCudnnfMHAScaleMaskSoftmaxDropoutCallTarget ||
+          instr->custom_call_target() == kCudnnfMHASoftmaxDropoutCallTarget);
+}
+
+StatusOr<bool> IsBatchedDot(const HloInstruction* gemm) {
+  GemmBackendConfig config;
+  TF_ASSIGN_OR_RETURN(config, gemm->backend_config<GemmBackendConfig>());
+  const DotDimensionNumbers& dot_dims = config.dot_dimension_numbers();
+  bool is_batch_dot = !dot_dims.lhs_batch_dimensions().empty() ||
+                      !dot_dims.rhs_batch_dimensions().empty();
+  return is_batch_dot;
+}
+
+bool IsBatchedMatmul(const HloInstruction* gemm) {
+  return IsCublasGemm(*gemm) && IsBatchedDot(gemm).value();
+}
+
+// Give this instruction a more useful name than "custom-call.42".
+Status SetName(HloModule* module, HloInstruction* fmha) {
+  if (fmha->custom_call_target() == kCudnnfMHABmmBmmCallTarget) {
+    module->SetAndUniquifyInstrName(fmha, "fmha-bmm-bmm");
```

Comment:
Nit: why not just use the existing string constant?

Reply:
Which string constant exactly? the custom-call target string?